### PR TITLE
[WIP] Rustup to rustc 1.15.0-nightly (0ed951993 2016-11-14)

### PR DESCRIPTION
**Pull request by mcarton**
:warning:  DO NOT MERGE!

This is compiling, but not working at all yet.

See https://github.com/rust-lang/rust/pull/37402, https://github.com/rust-lang/rust/pull/37408, https://github.com/rust-lang/rust/pull/37412, https://github.com/rust-lang/rust/pull/37688 and probably others.

### CODE REVIEW

File: tests/compile-fail/diverging_sub_expression.rs [hunk 1]
```diff
@@ -26,16 +25,16 @@ fn foobar() {
...
             9 => diverge(),
-            3 => (println!("moo"), diverge()), //~ ERROR sub-expression diverges
+            3 => true || diverge(), //~ ERROR sub-expression diverges
             10 => match 42 {
                 99 => return,
-                _ => ((), panic!("boo")),
+                _ => true || panic!("boo"),
             },
-            _ => (println!("boo"), break), //~ ERROR sub-expression diverges
+            _ => true || break, //~ ERROR sub-expression diverges
```
Comment (mcarton) [inline tests/compile-fail/diverging_sub_expression.rs]: It looks like rust is able to find some of those by itself now.

**Comment by mcarton** [comment]
/me just realized that tests are super-slow because he had `RUST_BACKTRACE=1`, not because of something he'd changed

**Comment by mcarton** [comment]
Well, tests pass locally now, but I want to review that later. There was some huge changes.

File: clippy_lints/src/utils/inspector.rs [hunk 2]
```diff
@@ -350,25 +350,25 @@ fn print_item(cx: &LateContext, item: &hir::Item) {
...
-            println!("enum definition: {:?}", cx.tcx.opt_lookup_item_type(did));
+            //println!("enum definition: {:?}", cx.tcx.opt_lookup_item_type(did));
         },
         hir::ItemStruct(..) => {
-            println!("struct definition: {:?}", cx.tcx.opt_lookup_item_type(did));
+            //println!("struct definition: {:?}", cx.tcx.opt_lookup_item_type(did));
         },
         hir::ItemUnion(..) => {
-            println!("union definition: {:?}", cx.tcx.opt_lookup_item_type(did));
+            //println!("union definition: {:?}", cx.tcx.opt_lookup_item_type(did));
```
Comment (mcarton) [inline clippy_lints/src/utils/inspector.rs]: Oops, forgot these, thanks dogfood :sweat_smile:

**Comment by oli-obk** [comment]
we kinda duplicated work... oh well...

File: clippy_lints/src/block_in_if_condition.rs [hunk 3]
```diff
@@ -55,17 +55,9 @@ struct ExVisitor<'v> {
...
-                if block.stmts.is_empty() {
-                    if let Some(ref ex) = block.expr {
-                        matches!(ex.node, ExprBlock(_))
-                    } else {
-                        false
-                    }
-                } else {
-                    true
-                }
+                matches!(expr.node, ExprBlock(_))
```
Comment (mcarton) [inline clippy_lints/src/block_in_if_condition.rs]: Note: add tests for that.

File: clippy_lints/src/escape.rs [hunk 4]
```diff
@@ -146,32 +145,35 @@ impl<'a, 'tcx: 'a+'gcx, 'gcx: 'a> Delegate<'tcx> for EscapeDelegate<'a, 'tcx, 'g
...
+                    if let Some(&Adjust::DerefRef { autoderefs, .. }) = self.tcx
                                                                             .tables
                                                                             .borrow()
                                                                             .adjustments
                                                                             .get(&self.tcx
-                                                                                      .map
-                                                                                      .get_parent_node(borrow_id)) {
-                        if adj.autoderefs <= 1 {
+                                                                            .map
+                                                                            .get_parent_node(borrow_id))
```
Comment (mcarton) [inline clippy_lints/src/escape.rs]: Fix indent here.

File: clippy_lints/src/map_clone.rs [hunk 5]
```diff
@@ -30,11 +30,8 @@ impl LateLintPass for Pass {
         if let ExprMethodCall(name, _, ref args) = expr.node {
             if name.node.as_str() == "map" && args.len() == 2 {
                 match args[1].node {
-                    ExprClosure(_, ref decl, ref blk, _) => {
+                    ExprClosure(_, ref decl, ref closure_expr, _) => {
                         if_let_chain! {[
-                            // just one expression in the closure
-                            blk.stmts.is_empty(),
-                            let Some(ref closure_expr) = blk.expr,
                             // nothing special in the argument, besides reference bindings
```
Comment (mcarton) [inline clippy_lints/src/map_clone.rs]: Note: Check that.

File: clippy_lints/src/utils/mod.rs [hunk 6]
```diff
@@ -429,7 +429,12 @@ pub fn get_enclosing_block<'c>(cx: &'c LateContext, node: NodeId) -> Option<&'c
     if let Some(node) = enclosing_node {
         match node {
             Node::NodeBlock(block) => Some(block),
-            Node::NodeItem(&Item { node: ItemFn(_, _, _, _, _, ref block), .. }) => Some(block),
+            Node::NodeItem(&Item { node: ItemFn(_, _, _, _, _, ref expr), .. }) => {
+                match expr.node {
+                    ExprBlock(ref block) => Some(block),
+                    _ => None,
+                }
+            }
```
Comment (mcarton) [inline clippy_lints/src/utils/mod.rs]: Note: do we ever need the enclosing block of `|| 1`?

**Comment by oli-obk** [comment]
> Well, tests pass locally now, but I want to review that later. There was some huge changes.

push it, I'll do a review round

File: clippy_lints/src/utils/inspector.rs (hunk 2 above)
Comment (oli-obk) [inline clippy_lints/src/utils/inspector.rs]: Just change it to item_type or remove the info entirely just leaving the "union definition" text

**Comment by oli-obk** [comment]
I took the liberty of appending to your PR

**Comment by mcarton** [comment]
> I took the liberty of appending to your PR

Thanks.
I'm going to publish it.

**Comment by mcarton** [comment]
Almost 0.0.:100:!

<sub><sup>
(although it's [probably not the 100th version](https://github.com/Manishearth/rust-clippy/blob/master/CHANGELOG.md#0052))
<sub><sup>

**Pull request merged**